src/main/scala/stravapr/RateLimiter.scala

```diff
@@ -20,6 +20,33 @@ import kiambogo.scrava.models.RateLimitException
 
 import scala.concurrent.duration._
 import scala.util.control.NonFatal
+import java.util._
+
+class StravaRateLimiter(isRateLimitExceeded : Throwable => Boolean) extends RateLimiter(isRateLimitExceeded) {
```

Comment:
We don't need to replicate so much of this logic.

I think we can have some sort of parameterizeble backup strategy.  For instance, we can have something an ADT like:

```scala
sealed trait BackoffStrategy
object BackoffStrategy {
  case class ExponentialBackoff(startBackoff: Duration, exponent: Double = 2.0) extends BackoffStrategy
  case class ConstantBackoff(backoff: Duration) extends BackoffStrategy
}
```

This way we wouldn't need a `StravaRateLimiter` and `RateLimiter` would receive a `BackoffStrategy`.  E.g.

```
object RateLimiter {
  ...

  def byException[T <: Throwable: ClassTag]: RateLimiter =
    new RateLimiter(exceptionCausedBy[T], BackoffStrategy.ExponentialBackoff(startBackoff = 1.minute, exponent = 2.0))

  def byExceptionForStrava[T <: Throwable: ClassTag]: RateLimiter =
    new RateLimiter(exceptionCausedBy[T], BackoffStrategy.ConstantBackoff(startBackoff = 15.minutes))
}
```

We could still have the nice message saying that we are waiting x time, and will resume at y.  Also, to print that message you do some math and manual formatting.  There are methods that can help you with that (e.g. take a look at `Instant` (where you can do math with durations) and `Duration`).

Reply:
the constantbackoff is not optimal as you will always wait for 15 minutes even thiugh you may only need 1 minute if the exception occurs let say at 03:14 and rate limit will be reset at 3:15 as strava API states.

so the constantbackoff must be renamed as the baxkoff is not constant at all.

I will try to provide a refactoring using a strategy pattern approach.

I agree for the flrmatting part.

Reply:
Yeah, it can be a `BackoffStrategy.StravaBackoff` or something like that :)